dusk-labyrinth.lic

```diff
@@ -7,74 +7,208 @@ custom_require.call %w[common common-items events]
+    ]
+
+    args = parse_args(arg_definitions)
+    $debug_mode_dl = UserVars.dusk_labyrinth_debug || args.debug || false
+
     @settings = get_settings
     @labyrinth_data = @settings.duskruin['labyrinth']
+
     loot_container = @labyrinth_data['loot_container']
     pet_noun = @labyrinth_data['pet_noun']
     redeem_scrip = @labyrinth_data['redeem_scrip']
+
+    # Map of directions to help know where we've come from and how to go back.
+    @reverse_direction_map = {
+      'e' => 'w',
+      'w' => 'e',
+      's' => 'n',
+      'n' => 's',
+      'ne' => 'sw',
+      'sw' => 'ne',
+      'nw' => 'se',
+      'se' => 'nw',
+      'east' => 'west',
+      'west' => 'east',
+      'south' => 'north',
+      'north' => 'south',
+      'northeast' => 'southwest',
+      'southwest' => 'northeast',
+      'northwest' => 'southeast',
+      'southeast' => 'northwest'
+    }
+
+    # Be notified when certain events happen, such as seeing a pet move or the maze reshuffle.
     Flags.add('five-minute-warning', 'You only have about 5 minutes left in the labyrinth!')
     Flags.add('done', 'A surly Dwarf stomps in and glowers at you in the dim gloom')
     Flags.add('no-pet', "A small #{pet_noun} scurries #{Regexp.new('(?<pet_dir>\w+)')}!")
     Flags.add('pet', "You notice a small #{pet_noun} scurrying around the area")
     Flags.add('caught', 'You pick them both up, claiming your new pet')
-    @prev_dir = nil
-    @dir_to_prev_dir = {
-      'northeast' => 'southwest',
-      'southwest' => 'northeast',
-      'northwest' => 'southeast',
-      'southeast' => 'northwest',
-      'north' => 'south',
-      'south' => 'north',
-      'east' => 'west',
-      'west' => 'east',
-      'up' => 'down', # up and down dont appear in the maze, but are used for testing
-      'down' => 'up'
-    }
+    Flags.add('maze-shuffled', 'no longer certain of your directions')
+
+    # Let's do this!
+    init_maze
     stow_hands
     main(loot_container) until done?
     redeem if redeem_scrip
   end
 
-  def redeem
-    while bput('get my bloodscrip', 'You get', 'What were') == 'You get'
-      bput('redeem my bloodscrip', 'You quickly pocket')
-    end
+  # Used at beginning of script and any time the maze shuffles.
+  def init_maze
+    echo("*** initializing map data") if $debug_mode_dl
+    # Maze data for the current room.
+    @current_room = nil
+    # Maze data for the previous room.
+    @previous_room = nil
+    # Direction we last moved to go from "previous_room" to "current_room".
+    @forward_move = nil
+    @reverse_move = nil
   end
 
+  # Main loop to check our room, search, or wander the maze.
+  # Repeatedly invoked until the maze event ends.
   def main(loot_container)
-    # search
-    @pause_timer = Time.now
-    pause 1 until Flags['no-pet'] || Flags['pet'] || Flags['caught']
+    if @previous_room && @previous_room['exits'][@forward_move]['count'] > 1 && !Flags['caught'] && !Flags['maze-shuffled'] && !Flags['five-minute-warning']
```

Comment:
Uh, not linked to this line, but this is what started my thinking. Would you be better off just having a boolean in your room nodes that tracks if you've visited the room before? It would be a lot more conceptually clear than asking 'have I traveled down the edge to this node from the previous node before. Also unless you're incrementing all neighbors from this room that travel to this room (or all their edges) it seems like this would consider a room untraveled if you arrived at it from a different neighbor than you had before? 

Honestly you could make new 'rooms' for every direction out of your current room that need them, then you don't need edge weights for graph transversal (all costs are identical anyways) and are just looking for the closest node that hasn't been visited yet.

Reply:
Very astute observations, and I agree with you. I got hung up on the idea of a "weighted" graph to influence which room to move to next when the core goal is just to track "have I been here before?" and if yes, immediately move to a not-yet-visited room rather than sit and redundantly wait for messaging that you won't find a pet here.

We're at the tail end of Duskruin, I won't have time to refactor and test before this year's event closes. How would you like to proceed -- leave this open and revisit next year or merge as-is or secret option 3?

Thanks again for taking time to provide great feedback; it's helping me a lot with learning Ruby and Lich scripting =)

Reply:
I don't play anymore so I'm not up on the scheduling, if dusk ruin is basically over and you've had a chance to test this so it's at least not worse then the old logic as far as repeating rooms I wouldn't be opposed to @rpherbig merging this in with the hope that next year you're still contributing and can take a fresh look at it. Alternatively it also doesn't hurt to just leave this PR open until it can be finished next year. I don't feel strongly so I'll leave it to you two to sort out.

Welcome to Lich and glad to have a new set of interested hands helping out.

Reply:
I'm inclined to merge this, then. I like better-but-not-perfect over waiting until next year.

Reply:
Thanks!

Reply:
Thanks!